### PR TITLE
restrict auth by list of allowed domains and emails

### PR DESCRIPTION
**Pull request by valeriykundas**

### CODE REVIEW

File: jetkit/api/auth/__init__.py [hunk 1]
```diff
@@ -82,15 +82,29 @@ def refresh_token():
...
     @blp.route("sign-up", methods=["POST"])
     @blp.response(user_schema)
     @blp.arguments(AuthRequest, as_kwargs=True)
     def sign_up(email: str, password: str):
         """Sign up with email and password. Possibly add other fields later."""
         cleaned_email = email.strip().lower()
-        existing_user: AuthModel = auth_model.query.filter_by(email=cleaned_email).one_or_none()
+
+        domain = email[email.index('@') + 1:]
+        allowed_by_domain = domain in allowed_domains
```
Comment (revmischa) [inline jetkit/api/auth/__init__.py]: shouldn't this check if allowed_domains/allowed_emails is set?

File: jetkit/api/auth/__init__.py [hunk 2]
```diff
@@ -82,15 +82,29 @@ def refresh_token():
...
+    # Since sign up can require not only email/password, separate this from core auth api
     @blp.route("sign-up", methods=["POST"])
     @blp.response(user_schema)
     @blp.arguments(AuthRequest, as_kwargs=True)
     def sign_up(email: str, password: str):
         """Sign up with email and password. Possibly add other fields later."""
         cleaned_email = email.strip().lower()
-        existing_user: AuthModel = auth_model.query.filter_by(email=cleaned_email).one_or_none()
+
+        domain = email[email.index('@') + 1:]
```
Comment (revmischa) [inline jetkit/api/auth/__init__.py]: what if `@` doesn't appear in the email?
parsing email addresses this way looks sketchy. you can use https://docs.python.org/3/library/email.utils.html#email.utils.parseaddr

File: jetkit/api/auth/__init__.py [hunk 3]
```diff
@@ -82,15 +82,29 @@ def refresh_token():
...
+        allowed_domains: Optional[List] = None,
+        allowed_emails: Optional[List] = None,
+):
+    # Since sign up can require not only email/password, separate this from core auth api
     @blp.route("sign-up", methods=["POST"])
     @blp.response(user_schema)
     @blp.arguments(AuthRequest, as_kwargs=True)
     def sign_up(email: str, password: str):
         """Sign up with email and password. Possibly add other fields later."""
         cleaned_email = email.strip().lower()
```
Comment (revmischa) [inline jetkit/api/auth/__init__.py]: consider https://pypi.org/project/email-normalize/ as well here

File: jetkit/api/auth/__init__.py (hunk 2 above)
Comment (valeriykundas) [reply jetkit/api/auth/__init__.py]: `AuthRequest` is a `marshmallow.Schema` that validates an email.
if the email does not contain `@`, an exception is thrown.
also, verified that each email contains exactly one `@`

`email.utils.parseaddr` does different job.
`utils.parseaddr('valerii <valerii@jetbridge.com>') == ('valerii', 'valerii@jetbridge.com')`
don't see how it helps there.

`utils.parseaddr('a') == ('', 'a') - it does not validate an email also.
`

File: jetkit/api/auth/__init__.py (hunk 3 above)
Comment (valeriykundas) [reply jetkit/api/auth/__init__.py]: added

File: jetkit/api/auth/__init__.py (hunk 2 above)
Comment (revmischa) [reply jetkit/api/auth/__init__.py]: ok

File: jetkit/api/auth/__init__.py (hunk 1 above)
Comment (valeriykundas) [reply jetkit/api/auth/__init__.py]: added checks in next commit

File: jetkit/test/test_auth.py [hunk 4]
```diff
@@ -87,3 +88,29 @@ def test_sign_up(client_unauthenticated, api_auth, client):
     )
 
     assert not sign_up_response == 200
+
+
+def test_validate_email():
+    test_email = "me@mail.com"
+
+    assert validate_email(test_email, allowed_domains=["mail.com"], allowed_emails=None)
```
Comment (gnudeb) [inline jetkit/test/test_auth.py]: Looks like a usecase for pytest's `parametrize` (https://docs.pytest.org/en/2.8.7/parametrize.html)

File: jetkit/api/auth/__init__.py [hunk 5]
```diff
@@ -82,15 +83,53 @@ def refresh_token():
         return {"access_token": create_access_token(identity=current_user)}
 
 
-def use_sign_up_api(auth_model: AuthModel, user_schema: Type[Schema] = UserSchema):
-    # Since sign up can require not only email/password, seperate this from core auth api
+def validate_email(
+        email: str,
+        allowed_domains: Union[str, Optional[List[str]]],
```
[the remark this answers is not in the document]
Comment (valeriykundas) [reply jetkit/api/auth/__init__.py]: fixed

File: jetkit/test/test_auth.py (hunk 4 above)
Comment (valeriykundas) [reply jetkit/test/test_auth.py]: nice. fixed

**Pull request merged**